packages/cli/src/util/get-latest-version/get-latest-worker.js

```diff
@@ -0,0 +1,189 @@
+/**
+ * This file is spawned in the background and checks npm for the latest version
+ * of the CLI, then writes the version to the cache file.
+ *
+ * NOTE: Since this file runs asynchronously in the background, it's possible
+ * for multiple instances of this file to be running at the same time leading
+ * to a race condition where the most recent instance will overwrite the
+ * previous cache file resetting the `notified` flag and cause the update
+ * notification to appear for multiple consequetive commands. Not the end of
+ * the world, but something to be aware of.
+ */
+
+const fetch = require('node-fetch');
+const fs = require('fs-extra');
+const path = require('path');
+const { Agent: HttpsAgent } = require('https');
+const { bold, gray, red } = require('chalk');
+const { format, inspect } = require('util');
+
+/**
+ * An simple output helper which accumulates error and debug log messages in
+ * memory for potential persistance to disk while immediately outputting errors
+ * and debug messages, when the `--debug` flag is set, to `stderr`.
+ */
+class WorkerOutput {
+  debugLog = [];
+  logFile = null;
+
+  constructor({ debug = true }) {
+    this.debugOutputEnabled = debug;
+  }
+
+  debug(...args) {
+    this.print('debug', args);
+  }
+
+  error(...args) {
+    this.print('error', args);
+  }
+
+  print(type, args) {
+    const str = format(
+      ...args.map(s => (typeof s === 'string' ? s : inspect(s)))
+    );
+    this.debugLog.push(`[${new Date().toISOString()}] [${type}] ${str}`);
+    if (type === 'debug' && this.debugOutputEnabled) {
+      console.error(
+        `${gray('>')} ${bold('[debug]')} ${gray(
+          `[${new Date().toISOString()}]`
+        )} ${str}`
+      );
+    } else if (type === 'error') {
+      console.error(`${red(`Error:`)} ${str}`);
+    }
+  }
+
+  setLogFile(file) {
+    // wire up the exit handler the first time the log file is set
+    if (this.logFile === null) {
+      process.on('exit', () => {
+        if (this.debugLog.length) {
+          fs.outputFileSync(this.logFile, this.debugLog.join('\n'));
+        }
+      });
+    }
+    this.logFile = file;
+  }
+}
+
+const output = new WorkerOutput({
+  // enable the debug logging if the `--debug` is set or if this worker script
+  // was directly executed
+  debug: process.argv.includes('--debug') || !process.connected,
+});
+
+process.on('unhandledRejection', err => {
+  output.error('Exiting worker due to unhandled rejection:', err);
+  process.exit(1);
+});
+
+const defaultUpdateCheckInterval = 1000 * 60 * 60 * 24 * 7; // 1 week
+
+// this timer will prevent this worker process from running longer than 10s
+const timer = setTimeout(() => {
+  output.error('Worker timed out after 10 seconds');
+  process.exit(1);
```

Comment:
I think this needs to clean up the lock file. Otherwise the user will never be able to update again.

I'm wondering if we even need a lock file at all 🤔

Reply:
This `setTimeout()` exists in the event the parent never sends the message, thus no lock file will be created to be cleaned up. The lock file name is derived from the cache file name, which is the package name plus the dist tag. Those two bits of info are not available at the time the timer is created.

We could rename `timer` to `waitForParentTimer`, then clear it once it gets the message to start, then create an `AbortController` to make sure the `fetch()` returns in a reasonable time.

Lock files are only 4KB on disk and I have yet to come across a scenario where the lock file wasn't cleaned up. I'm sure it will happen at some point and that's why I added the stale pid check in `isRunning()`.

As for needing the lock file, I did some testing and it's possible to run `vercel` commands fast enough that one worker is clobbering the cache file with stale data (e.g. the `notified` flag being toggled). Since the access to the cache file should be atomic, a lock file will synchronize multiple processes ensuring a predictable state. Obviously we're not building a database, so if you feel the lock file, or the whole IPC design for that matter, is overkill, I'd be happy to refactor the code.